File: node-martian-ddd/src/app/mission.js
"use strict";
const { Robot } = require("../domain/robot");
const { ScentRepository } = require("../domain/scents");
const { execute } = require("../domain/navigator");
const {
  parseGrid,
  parseRobot,
  normalizeInstr,
  splitNonEmpty,
} = require("./parse");

function runMission(input) {
  const lines = splitNonEmpty(input);
  if (lines.length === 0) return "\n";
  const { maxX, maxY } = parseGrid(lines[0]);
  const grid = { maxX, maxY };
  const scents = new ScentRepository();
  const out = [];
  for (let i = 1; i + 1 < lines.length; i += 2) {
    const { x, y, o } = parseRobot(lines[i]);
    const instr = normalizeInstr(lines[i + 1]);
    const robot = new Robot({ x, y }, o);
    execute(scents, grid, robot, instr);
    out.push(
      `${robot.position.x} ${robot.position.y} ${robot.orientation}${robot.lost ? " LOST" : ""}`,
    );
  }
  return out.join("\n") + "\n\n";
}

module.exports = { runMission };

File: langs/ocaml-ddd/mission.ml
open Grid
open Scents
open Robot
open Navigator
open Parsing
open Orientation

let read_all ic =
  let rec aux acc =
    try let s = input_line ic |> String.trim in aux (if s<>"" then s::acc else acc) with End_of_file -> List.rev acc
  in aux []

let run ic oc =
  match read_all ic with
  | [] -> output_string oc "\n"
  | header::rest ->
      let mx,my = parse_grid header in
      let grid = { max_x = mx; max_y = my } in
      let rec loop lines scents =
        match lines with
        | meta :: instr :: tail ->
            let pos,o = parse_robot meta in
            let r = create pos o in
            let instr = normalize instr in
            execute (module In_memory : REPO) grid r instr;
            Printf.fprintf oc "%d %d %s%s\n" r.position.x r.position.y (to_string r.orientation) (if r.lost then " LOST" else "");
            loop tail scents
        | _ -> ()
      in
      loop rest (); output_char oc '\n'
